library/core/src/ffi/c_str.rs

```diff
@@ -735,3 +758,69 @@ const unsafe fn const_strlen(ptr: *const c_char) -> usize {
         intrinsics::const_eval_select((ptr,), strlen_ct, strlen_rt)
     }
 }
+
+/// An iterator over the bytes of a [`CStr`], without the nul terminator.
+///
+/// This struct is created by the [`bytes`] method on [`CStr`].
+/// See its documentation for more.
+///
+/// [`bytes`]: CStr::bytes
+#[must_use = "iterators are lazy and do nothing unless consumed"]
+#[unstable(feature = "cstr_bytes", issue = "112115")]
+#[derive(Clone, Debug)]
+pub struct Bytes<'a> {
+    // since we know the string is nul-terminated, we only need one pointer
+    ptr: NonNull<u8>,
+    phantom: PhantomData<&'a u8>,
+}
+impl<'a> Bytes<'a> {
+    #[inline]
+    fn new(s: &'a CStr) -> Self {
+        Self {
+            // SAFETY: Because we have a valid reference to the string, we know
+            // that its pointer is non-null.
+            ptr: unsafe { NonNull::new_unchecked(s.as_ptr() as *const u8 as *mut u8) },
```

Comment:
This can be built from a safe reference, although it won't be as cheap if `CStr` goes thin.
```suggestion
            ptr: NonNull::from(s.to_bytes_with_nul()).as_non_null_pointer(),
```

Reply:
Let me take a look to see if I can make a version that will remain cheap in that case.

Reply:
Yup, turns out that `cast` exists for `NonNull` too. So, I'll just do that instead of having to go through `to_bytes_with_nul`.